# Hands on: Data Preparation


### Overview

- [The Data](#chi)</a>

- [Import the file titanicData.csv to a dataframe](#chapi_1)</a>

- [Obtain a summary of the data.](#chapi_2)</a>

    - [Distribution of numerical features](#chapi_2_1)</a>
    
    - [Distribution of categorical features](#chapi_2_2)</a>

-----------------

[1. Find out how many missing values are represented](#ch_1)</a>

[2. Handling missings](#ch_2)</a>

[3. Scaling/normalization](#ch_3)</a>

[4. Discretization](#ch_4)</a>

[5. Binarization](#ch_5)</a>

[6. Feature engineering ](#ch_6)</a>

In [84]:
# to execute and return the results of executions
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


# The Data <a name="chi"></a>
### The titanic dataset contains information on the survival status of individual passengers on the Titanic 

**Pass_class** - Ticket class (1 = 1st, 2 = 2nd, 3 = 3rd)
    
**name**	

**sex**	

**age**	

**sibsp** - Number of Siblings/Spouses Aboard

**parch** - Number of Parents/Children Aboard	

**ticket** - Ticket number	

**fare** - Passenger fare	

**cabin**	- Cabin number	

**embarked** - Port of Embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)
    
**survived** - Survival status (0 = No, 1 = Yes)


## Import the file titanicData.csv to a dataframe (csv file is available in eLearning). <a name="chapi_1"></a>

In [85]:
import numpy as np
import pandas as pd
df = pd.read_csv("titanicData.csv", sep=';')
df

Pass_class                                             name     sex  \
0              1                    Allen, Miss. Elisabeth Walton  female   
1              1                   Allison, Master. Hudson Trevor    male   
2              1                     Allison, Miss. Helen Loraine  female   
3              1             Allison, Mr. Hudson Joshua Creighton    male   
4              1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
...          ...                                              ...     ...   
1304           3                             Zabour, Miss. Hileni  female   
1305           3                            Zabour, Miss. Thamine  female   
1306           3                        Zakarian, Mr. Mapriededer    male   
1307           3                              Zakarian, Mr. Ortin    male   
1308           3                               Zimmerman, Mr. Leo    male   

          age  sibsp  parch  ticket      fare    cabin embarked  survived  
0     29.0000      0      0   24160  211.3375       B5        S         1  
1      0.9167      1      2  113781  151.5500  C22 C26        S         1  
2      2.0000      1      2  113781  151.5500  C22 C26        S         0  
3     30.0000      1      2  113781  151.5500  C22 C26        S         0  
4     25.0000      1      2  113781  151.5500  C22 C26        S         0  
...       ...    ...    ...     ...       ...      ...      ...       ...  
1304  14.5000      1      0    2665   14.4542      NaN        C         0  
1305      NaN      1      0    2665   14.4542      NaN        C         0  
1306  26.5000      0      0    2656    7.2250      NaN        C         0  
1307  27.0000      0      0    2670    7.2250      NaN        C         0  
1308  29.0000      0      0  315082    7.8750      NaN        S         0  

[1309 rows x 11 columns]

## Obtain a summary of the data  <a name="chapi_2"></a>

In [86]:
df.describe()

Pass_class          age        sibsp        parch         fare  \
count  1309.000000  1046.000000  1309.000000  1309.000000  1308.000000   
mean      2.294882    29.881135     0.498854     0.385027    33.295479   
std       0.837836    14.413500     1.041658     0.865560    51.758668   
min       1.000000     0.166700     0.000000     0.000000     0.000000   
25%       2.000000    21.000000     0.000000     0.000000     7.895800   
50%       3.000000    28.000000     0.000000     0.000000    14.454200   
75%       3.000000    39.000000     1.000000     0.000000    31.275000   
max       3.000000    80.000000     8.000000     9.000000   512.329200   

          survived  
count  1309.000000  
mean      0.381971  
std       0.486055  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000

### Distribution of numerical features  <a name="chapi_2_1"></a>

This helps us determine, among other early insights, how representative is the training dataset of the actual problem domain.

    Total samples 1309
    Survived is a categorical feature with 0 or 1 values
    Around 38% samples survived representative of the actual survival rate at 32%
    Most passengers (> 75%) did not travel with parents or children
    Nearly 30% of the passengers had siblings and/or spouse aboard
    Fares varied significantly with few passengers (<25%) paying as high as $512
    Few elderly passengers (<25%) within age range 65-80



In [87]:
df.describe(include='object')

name   sex    ticket        cabin embarked
count                   1309  1309      1309          295     1307
unique                  1307     2       929          186        3
top     Connolly, Miss. Kate  male  CA. 2343  C23 C25 C27        S
freq                       2   843        11            6      914

### Distribution of categorical features <a name="chapi_2_2"></a>

    Names are not unique across the dataset (count=1309; unique=1308) - 2 duplicates
    Sex variable as two possible values with 843/1309% male (top=male, freq=843/count=1309)
    Cabin values have several dupicates across samples. Alternatively several passengers shared a cabin
    Embarked takes three possible values. S port used by most passengers (top=S)
    Ticket feature has high ratio of duplicate values (unique=929) - family tickets?


## 1.	Find out how many missing values are represented <a name="ch1"></a>

In [88]:
# get the number of missing data points per column
missing_values_count = df.isnull().sum()
print('Missing data objects per feature')
missing_values_count
print('-'*80)

# how many total missing values do we have?
total_missing = missing_values_count.sum() # sum(missing_values_count)
print('Total number of missing values in the dataset = ', total_missing)
NumberObjects = len(df.index)
print('Size of the dataset = ', NumberObjects)
print('-'*80)

# percent of data that is missing
total_cells = np.product(df.shape)
percent_missing = (total_missing/total_cells) * 100
print('percentage of total missing values in the dataset = ', percent_missing)

Missing data objects per feature


Pass_class       0
name             0
sex              0
age            263
sibsp            0
parch            0
ticket           0
fare             1
cabin         1014
embarked         2
survived         0
dtype: int64

--------------------------------------------------------------------------------
Total number of missing values in the dataset =  1280
Size of the dataset =  1309
--------------------------------------------------------------------------------
percentage of total missing values in the dataset =  8.889506215709424


## 2. Handling missings <a name="ch2"></a>

In [89]:
#Drop rows with missing values at specific columns
df['age'].fillna(df.groupby(['sex','Pass_class'])['age'].transform('mean'),inplace=True)
df.dropna(subset=['fare', 'embarked'], inplace = True)
df.drop('cabin', axis=1, inplace = True)

df.isnull().sum()

Pass_class    0
name          0
sex           0
age           0
sibsp         0
parch         0
ticket        0
fare          0
embarked      0
survived      0
dtype: int64

## 3. Scaling/normalization <a name="ch3"></a>

### 3.1. Standardize `fare`

***The following code standardizes by fitting and transforming at once.***

**If dealing with trainind and testing data sets, in order to avoid 'Data Leakage', the fit should be performed only in the training test, and then, transform should apply to both.**


In [90]:
from sklearn.preprocessing import StandardScaler
standard_scaler = StandardScaler()
fare_scaled = pd.DataFrame(standard_scaler.fit_transform(pd.DataFrame(df.fare.values)))
df['fare_std'] = pd.DataFrame(fare_scaled)
df['fare_std'] 
POS= df.columns.get_loc('fare') + 1
df.insert(POS, 'fare_std', df['fare_std'])


0       3.442063
1       2.286663
2       2.286663
3       2.286663
4       2.286663
          ...   
1304   -0.502433
1305   -0.489871
1306         NaN
1307         NaN
1308         NaN
Name: fare_std, Length: 1306, dtype: float64

## 3.1. minMax scaling `fare`

In [91]:
from sklearn.preprocessing import MinMaxScaler
min_max_scaler = MinMaxScaler()
fare_scaled = pd.DataFrame(min_max_scaler.fit_transform(pd.DataFrame(df.fare.values)))
df['fare_minMax'] = pd.DataFrame(fare_scaled)
df['fare_minMax'] 
POS= df.columns.get_loc('fare') + 2
df.insert(POS, 'fare_minMax', df['fare_minMax'])

0       0.412503
1       0.295806
2       0.295806
3       0.295806
4       0.295806
          ...   
1304    0.014102
1305    0.015371
1306         NaN
1307         NaN
1308         NaN
Name: fare_minMax, Length: 1306, dtype: float64

## 4. Discretization <a name="ch4"></a>

### 4.1. Create age intervals

In [92]:
NumbQuant = 3
df['ageInt'], cut_intervals = pd.qcut(df.age, NumbQuant, retbins=True)

## Question:
- **Analyse the relation of `ageInt` with `survived`.**
- **Investigate different number of quantiles (ranging from 3 to 6). Which value shows better relation with `survived`?**

In [93]:
print('Here is an example of code to allow studying the relation of ageInt and survived:')
#Investigate different number of quantiles (ranging from 3 to 6). Which value shows better relation with `survived`?
# place your code here!
from sklearn.preprocessing import MinMaxScaler
#range from 3 to 6 quantiles
for i in range(3,7):
    df['ageInt'], cut_intervals = pd.qcut(df.age, i, retbins=True)
    print('Number of quantiles = ', i)
    print(df.groupby('ageInt')['survived'].mean())
    ctab=pd.crosstab(df['ageInt'], df['survived'])
    ctab
    print('-'*80)




    



Here is an example of code to allow studying the relation of ageInt and survived:
Number of quantiles =  3
ageInt
(0.166, 24.0]    0.447034
(24.0, 32.0]     0.278049
(32.0, 80.0]     0.408019
Name: survived, dtype: float64


survived         0    1
ageInt                 
(0.166, 24.0]  261  211
(24.0, 32.0]   296  114
(32.0, 80.0]   251  173

--------------------------------------------------------------------------------
Number of quantiles =  4
ageInt
(0.166, 22.0]     0.435435
(22.0, 26.0]      0.298851
(26.0, 36.375]    0.399329
(36.375, 80.0]    0.397554
Name: survived, dtype: float64


survived          0    1
ageInt                  
(0.166, 22.0]   188  145
(22.0, 26.0]    244  104
(26.0, 36.375]  179  119
(36.375, 80.0]  197  130

--------------------------------------------------------------------------------
Number of quantiles =  5
ageInt
(0.166, 21.0]     0.431034
(21.0, 25.962]    0.313019
(25.962, 30.0]    0.366864
(30.0, 40.0]      0.433476
(40.0, 80.0]      0.383399
Name: survived, dtype: float64


survived          0    1
ageInt                  
(0.166, 21.0]   165  125
(21.0, 25.962]  248  113
(25.962, 30.0]  107   62
(30.0, 40.0]    132  101
(40.0, 80.0]    156   97

--------------------------------------------------------------------------------
Number of quantiles =  6
ageInt
(0.166, 19.0]     0.471111
(19.0, 24.0]      0.425101
(24.0, 26.0]      0.181818
(26.0, 32.0]      0.378109
(32.0, 41.029]    0.403756
(41.029, 80.0]    0.412322
Name: survived, dtype: float64


survived          0    1
ageInt                  
(0.166, 19.0]   119  106
(19.0, 24.0]    142  105
(24.0, 26.0]    171   38
(26.0, 32.0]    125   76
(32.0, 41.029]  127   86
(41.029, 80.0]  124   87

--------------------------------------------------------------------------------


## 5. Binarization <a name="ch5"></a>
**Converting a categorical feature into numeric**

### 5.1. Convert `ageInt` into a numeric feature `ageInt_num`

In [94]:
df.ageInt.unique()
NumbQuant = 3
cut_intervals

[(26.0, 32.0], (0.166, 19.0], (24.0, 26.0], (41.029, 80.0], (32.0, 41.029], (19.0, 24.0]]
Categories (6, interval[float64, right]): [(0.166, 19.0] < (19.0, 24.0] < (24.0, 26.0] < (26.0, 32.0] < (32.0, 41.029] < (41.029, 80.0]]

array([ 0.1667    , 19.        , 24.        , 26.        , 32.        ,
       41.02924967, 80.        ])

In [95]:
df.loc[df.age <= cut_intervals[0], 'ageInt_num'] = 0
ind = 0
for k in reversed(cut_intervals[1:-1]):
    ind = ind+1
    df.loc[(df.age > k) & (df.age > k+1), 'ageInt_num'] = ind


**remove `ageInt`**

In [96]:
# remove ageInt

df.drop('ageInt', axis=1, inplace = True)


### 5.2. Convert `sex` into a numeric feature `gender`

___________________
Now we can convert features which contain strings to numerical values. This is required by most model algorithms. Doing so will also help us in achieving the feature completing goal.

Let us start by converting Sex feature to a new feature called Gender where female=1 and male=0.

In [97]:
df['sex'] = df['sex'].map( {'female': 1, 'male': 0} ).astype(int)
df['sex']

0       1
1       0
2       1
3       0
4       1
       ..
1304    1
1305    1
1306    0
1307    0
1308    0
Name: sex, Length: 1306, dtype: int32

### 5.3. Convert `embarked` into a numeric feature `pe_num`

In [98]:
#place your code here!
df['embarked'] 
df['embarked'] = df['embarked'].map( {'C': 0, 'Q': 1, 'S': 2} ).astype(int)
df['embarked']



0       S
1       S
2       S
3       S
4       S
       ..
1304    C
1305    C
1306    C
1307    C
1308    S
Name: embarked, Length: 1306, dtype: object

0       2
1       2
2       2
3       2
4       2
       ..
1304    0
1305    0
1306    0
1307    0
1308    2
Name: embarked, Length: 1306, dtype: int32

**remove `embarked`**

In [99]:
#place your code here!
#remove embarked
df.drop('embarked', axis=1, inplace = True)


## 6. Feature engineering <a name="ch6"></a>
**creating new feature from existing ones**

### 6.1. Create `title` feature
**Analyze if `Name` can be engineered to extract titles and test relation between titles and survival, before dropping `Name` and `PassengerId`:**
- after the creation of the feature, analyse if `title` is related with `survived`´.

In [100]:
df['title'] = [x.split(',')[1].split('.')[0].strip() for x in df['name']]

df['title'].unique()
pd.crosstab(df['title'], df['sex'])

array(['Miss', 'Master', 'Mr', 'Mrs', 'Col', 'Mme', 'Dr', 'Major', 'Capt',
       'Lady', 'Sir', 'Mlle', 'Dona', 'Jonkheer', 'the Countess', 'Don',
       'Rev', 'Ms'], dtype=object)

sex             0    1
title                 
Capt            1    0
Col             4    0
Don             1    0
Dona            0    1
Dr              7    1
Jonkheer        1    0
Lady            0    1
Major           2    0
Master         61    0
Miss            0  259
Mlle            0    2
Mme             0    1
Mr            756    0
Mrs             0  196
Ms              0    2
Rev             8    0
Sir             1    0
the Countess    0    1

In [101]:
#joint as others the less representative titles
df['title'] = np.where(df['title'].isin(['Miss', 'Mr', 'Mrs', 'Mme', 'Lady', 'Ms', 'Mlle', 'Dona', 'Don']), df['title'], 'Other')
df['title'] = df['title'].replace(['Ms', 'Lady', 'Mlle'],'Miss')
df['title'] = df['title'].replace(['Mme', 'Dona'],'Mrs')
df['title'] = df['title'].replace('Don','Mr')
df['title'].unique()
df['title'].value_counts()

array(['Miss', 'Other', 'Mr', 'Mrs'], dtype=object)

Mr       757
Miss     264
Mrs      198
Other     87
Name: title, dtype: int64

- **confirm if `Title` is related with `survived`**

In [108]:
# place your code here!
#confirm if title is related with survived
df.groupby('title')['survived'].mean()
#or use crosstab
ctab=pd.crosstab(df['title'], df['survived'])
ctab





title
Miss     0.678030
Mr       0.162483
Mrs      0.787879
Other    0.459770
Name: survived, dtype: float64

survived    0    1
title             
Miss       85  179
Mr        634  123
Mrs        42  156
Other      47   40

### What to conclude? 

#### Drop`name`

In [103]:
# place your code here!
#remove name
df.drop('name', axis=1, inplace = True)

### 6.2. Create `FamilySize` feature
**Creation of the new feature `FamilySize` which combines `sibsp`and `parch`.**

In [104]:
# place you code here!
#create FamilySize feature as a combination of SibSp and Parch
df['familySize'] = df['sibsp'] + df['parch'] + 1
df['familySize'].value_counts()


1     787
2     235
3     159
4      43
6      25
5      22
7      16
11     11
8       8
Name: familySize, dtype: int64

- **confirm if `Title` is related with `survived`**

In [109]:
# place you code here!
#confirm if FamilySize is related with survived
df.groupby('familySize')['survived'].mean()
#or use crosstab
ctab=pd.crosstab(df['familySize'], df['survived'])
ctab


familySize
1     0.301144
2     0.536170
3     0.566038
4     0.697674
5     0.272727
6     0.200000
7     0.250000
8     0.000000
11    0.000000
Name: survived, dtype: float64

survived      0    1
familySize          
1           550  237
2           109  126
3            69   90
4            13   30
5            16    6
6            20    5
7            12    4
8             8    0
11           11    0

**Do not drop `sibsp`and `parch`. We will keep them!**

### 6.3. Create `isAlone` feature
**Creation of the new feature `isAlone` based on `familySize`.**

In [106]:
df

Pass_class  sex        age  sibsp  parch  ticket      fare  survived  \
0              1    1  29.000000      0      0   24160  211.3375         1   
1              1    0   0.916700      1      2  113781  151.5500         1   
2              1    1   2.000000      1      2  113781  151.5500         0   
3              1    0  30.000000      1      2  113781  151.5500         0   
4              1    1  25.000000      1      2  113781  151.5500         0   
...          ...  ...        ...    ...    ...     ...       ...       ...   
1304           3    1  14.500000      1      0    2665   14.4542         0   
1305           3    1  22.185307      1      0    2665   14.4542         0   
1306           3    0  26.500000      0      0    2656    7.2250         0   
1307           3    0  27.000000      0      0    2670    7.2250         0   
1308           3    0  29.000000      0      0  315082    7.8750         0   

      fare_std  fare_minMax  ageInt_num  title  familySize  
0     3.442063     0.412503         5.0   Miss           1  
1     2.286663     0.295806         NaN  Other           4  
2     2.286663     0.295806         NaN   Miss           4  
3     2.286663     0.295806         5.0     Mr           4  
4     2.286663     0.295806         5.0    Mrs           4  
...        ...          ...         ...    ...         ...  
1304 -0.502433     0.014102         NaN   Miss           2  
1305 -0.489871     0.015371         5.0   Miss           2  
1306       NaN          NaN         5.0     Mr           1  
1307       NaN          NaN         5.0     Mr           1  
1308       NaN          NaN         5.0     Mr           1  

[1306 rows x 13 columns]

In [110]:
df['isAlone'] = 0*df['familySize'] 
df.loc[(df.familySize == 1), 'isAlone'] = 1
df['isAlone'].value_counts()

1    787
0    519
Name: isAlone, dtype: int64